api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/PregnancyProgramEnrollmentAction.java

```diff
@@ -0,0 +1,70 @@
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.openmrs.Encounter;
+import org.openmrs.Patient;
+import org.openmrs.PatientProgram;
+import org.openmrs.Program;
+import org.openmrs.ProgramWorkflowState;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.htmlformentry.CustomFormSubmissionAction;
+import org.openmrs.module.htmlformentry.FormEntryContext;
+import org.openmrs.module.htmlformentry.FormEntrySession;
+import org.openmrs.module.pihcore.SierraLeoneConfigConstants;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.openmrs.module.pihcore.htmlformentry.action.util.PregnancyProgramActionUtils.endEnrollment;
+import static org.openmrs.module.pihcore.htmlformentry.action.util.PregnancyProgramActionUtils.enrollInPregnancyProgram;
+import static org.openmrs.module.pihcore.htmlformentry.action.util.PregnancyProgramActionUtils.getTypeOfTreatmentStateOnDate;
+
+/**
+ * Does the following when a *new* form is submitted:
+ * - If the Patient is not enrolled in the Pregnancy Program at the time of the encounter, enrolls the Patient in the Pregnancy Program with the state of Antenatal
+ * - If the Patient is already enrolled in the Pregnancy Program with the state of postpartum, end the current enrollment and enroll the Patient in the Pregnancy Program with the state of Antenatal
+ * Intended to be added to the following forms:
+ * - ANC Initial
+ * - ANC Followup
+ * - Labour Progress
+ * NOTE: this only works for the *Sierra Leone* implementation of the Pregnancy Program, and relies on constants from SierraLeoneConfigConstants (the generic Pregnancy Program has slightly different states)
+ */
+public class PregnancyProgramEnrollmentAction implements CustomFormSubmissionAction {
+
+    protected final Log log = LogFactory.getLog(getClass());
+
+    @Override
+    public void applyAction(FormEntrySession formEntrySession) {
+        // only do anything if we are entering (not editing) a form
+        if (formEntrySession.getContext().getMode().equals(FormEntryContext.Mode.ENTER)) {
+            Program pregnancyProgram = Context.getProgramWorkflowService().getProgramByUuid(SierraLeoneConfigConstants.PROGRAM_PREGNANCY_UUID);
+            ProgramWorkflowState antenatalState = Context.getProgramWorkflowService().getStateByUuid(SierraLeoneConfigConstants.PROGRAMWORKFLOW_PREGNANCYPROGRAMTYPEOFTREATMENT_STATE_ANTENATAL_UUID);
+
+            Patient patient = formEntrySession.getPatient();
+            Encounter encounter = formEntrySession.getEncounter();
+            // get all patient programs completed on or after encounter date (or not yet completed) and sort by date enrolled
+            List<PatientProgram> patientPregnancyPrograms = Context.getProgramWorkflowService().getPatientPrograms(patient, pregnancyProgram, null, null, encounter.getEncounterDatetime(), null, false);
```

Comment:
@mogoodrich , how about the programs whose enrollment date is prior to the encounter date, and the dateCompleted is null? Would this query return those programs as well?

Reply:
From the javadoc, I believe the one date parameter I set should "if supplied will limit PatientPrograms to those completed on or after this Date OR not yet completed"